src/ORM/DB.php

```diff
@@ -109,6 +126,124 @@ public static function get_conn($name = 'default')
+     * For instance disable when running unit-tests
+     */
+    public static function setCanUseReplicas(bool $canUseReplicas): void
+    {
+        DB::$canUseReplicas = $canUseReplicas;
+    }
+
+    /**
+     * Set the flag to only use the primary database connection for the current request
+     * meaning that replia connections will no longer be used
+     *
+     * This intentioally does not have a parameter to set this back to false, as this it to prevent
+     * accidentally attempting writing to a replica, or reading from an out of date replica
+     * after a write
+     */
+    public static function setMustUsePrimary(): void
+    {
+        DB::$mustUsePrimary = true;
+    }
+
+    /**
+     * Only use the primary database connection when calling $callback
+     * Use this when doing non-mutable queries on the primary database where querying
+     * an out of sync replica could cause issues
+     * There's no need to use this with mutable queries, or after calling a mutable query
+     * as the primary database connection will be automatically used
+     */
+    public static function withPrimary(callable $callback): mixed
+    {
+        DB::$withPrimaryCount++;
+        $result = $callback();
+        DB::$withPrimaryCount--;
+        return $result;
+    }
+
+    /**
+     * Get the name of the database connection to use for the given SQL query
+     * The 'default' connection can be either the primary or a replica connection if configured
+     */
+    private static function getDefaultConnectionName(string $sql = ''): string
+    {
+        if (DB::$mustUsePrimary || DB::$withPrimaryCount > 0 || !DB::$canUseReplicas || !DB::hasReplicaConfig()) {
+            return 'primary';
+        }
+        if ($sql) {
+            $dbClass = DB::getConfig('default')['type'];
+            // This must use getServiceSpec() and not Injector::get/create() followed by
+            // getConnector() as this can remove the dbConn from a different connection
+            // under edge case conditions
+            $dbSpec = Injector::inst()->getServiceSpec($dbClass);
+            $connectorService = $dbSpec['properties']['connector'];
+            $connector = Injector::inst()->convertServiceProperty($connectorService);
+            if ($connector->isQueryMutable($sql)) {
+                DB::$mustUsePrimary = true;
+                return 'primary';
+            }
+        }
+        if (DB::$randomReplicaConfigKey) {
+            return DB::$randomReplicaConfigKey;
+        }
+        $name = DB::getRandomReplicaConfigKey();
+        DB::$randomReplicaConfigKey = $name;
+        return $name;
+    }
+
+    /**
+     * Get a random replica database configuration key from the available replica configurations
+     * The replica choosen will be used for the rest of the request, unless the primary connection
+     * is forced
+     */
+    private static function getRandomReplicaConfigKey(): string
```

Comment:
Move the private method to the bottom of the class per https://docs.silverstripe.org/en/5/contributing/coding_conventions/php_coding_conventions/#class-member-ordering

Reply:
Done